File: tools/csharp-roslyn-host/Rules/Style/ExtensionKeywordConflict.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace TrueCourse.RoslynHost;

/// <summary>
/// A declared identifier literally named <c>extension</c>. C# 14 introduces
/// <c>extension</c> as a contextual keyword (extension blocks); an unescaped
/// identifier of that name will collide and must be written <c>@extension</c>.
/// We flag only declaration sites (types, members, parameters, locals) so a use of
/// a correctly-escaped name elsewhere is unaffected.
/// </summary>
internal sealed class ExtensionKeywordConflict : ISemanticRule
{
    public string RuleKey => "style/deterministic/extension-keyword-conflict";

    public IEnumerable<Violation> Analyze(SemanticModel model, SyntaxTree tree)
    {
        foreach (var token in tree.GetRoot().DescendantTokens())
        {
            if (!token.IsKind(SyntaxKind.IdentifierToken)) continue;
            // ValueText strips the leading @, so `@extension` is filtered out by
            // checking the raw token text still carries the escape.
            if (token.ValueText != "extension") continue;
            if (token.Text.StartsWith('@')) continue;
            if (!IsDeclarationIdentifier(token)) continue;

            var pos = token.GetLocation().GetLineSpan().StartLinePosition;
            yield return new Violation(
                RuleKey, tree.FilePath, pos.Line + 1, pos.Character + 1,
                "Identifier 'extension' collides with the C# 14 contextual keyword — rename it or escape it as '@extension'.");
        }
    }

    private static bool IsDeclarationIdentifier(SyntaxToken token)
    {
        var parent = token.Parent;
        return parent switch
        {
            BaseTypeDeclarationSyntax t => t.Identifier == token,
            DelegateDeclarationSyntax d => d.Identifier == token,
            MethodDeclarationSyntax m => m.Identifier == token,
            PropertyDeclarationSyntax p => p.Identifier == token,
            EventDeclarationSyntax e => e.Identifier == token,
            VariableDeclaratorSyntax v => v.Identifier == token,
            ParameterSyntax pa => pa.Identifier == token,
            EnumMemberDeclarationSyntax em => em.Identifier == token,
            SingleVariableDesignationSyntax sv => sv.Identifier == token,
            ForEachStatementSyntax fe => fe.Identifier == token,
            TypeParameterSyntax tp => tp.Identifier == token,
            LocalFunctionStatementSyntax lf => lf.Identifier == token,
            _ => false,
        };
    }
}

File: packages/analyzer/src/rules/style/visitors/csharp/extension-keyword-conflict.ts
import type { CodeRuleVisitor } from '../../../types.js'
import { makeViolation } from '../../../types.js'

// Declaration sites whose declared name can collide with the contextual keyword.
const DECLARATION_TYPES = [
  'class_declaration', 'struct_declaration', 'interface_declaration', 'record_declaration',
  'enum_declaration', 'delegate_declaration', 'method_declaration', 'property_declaration',
  'event_declaration', 'variable_declarator', 'parameter', 'enum_member_declaration',
  'type_parameter', 'local_function_statement',
]

/**
 * A declared identifier literally named `extension`. C# 14 introduces `extension`
 * as a contextual keyword (extension blocks); an unescaped identifier of that name
 * collides and must be written `@extension`. Only declaration sites are flagged —
 * an already-escaped `@extension` carries the escape in its node text and so is
 * never matched.
 */
export const csharpExtensionKeywordConflictVisitor: CodeRuleVisitor = {
  ruleKey: 'style/deterministic/extension-keyword-conflict',
  languages: ['csharp'],
  nodeTypes: DECLARATION_TYPES,
  visit(node, filePath, sourceCode) {
    const name = node.childForFieldName('name')
    if (!name || name.text !== 'extension') return null

    return makeViolation(
      this.ruleKey, name, filePath, 'low',
      "Identifier collides with the 'extension' contextual keyword",
      "Identifier 'extension' collides with the C# 14 contextual keyword — rename it or escape it as '@extension'.",
      sourceCode,
      "Rename the identifier, or escape it as '@extension'.",
    )
  },
}
